@joseeroman @dalcinl I get build errors with slepc4py that I don't understand.

```     40     In file included from src/SLEPc.c:3:
     41     src/slepc4py.SLEPc.c: In function '__Pyx__ExceptionSave':
  >> 42     src/slepc4py.SLEPc.c:68575:21: error: 'PyThreadState' {aka 'struct _ts'} has no member named 'exc_type'; did you mean 'curexc_type'?
     43     68575 |     *type = tstate->exc_type;
     44           |                     ^~~~~~~~
     45           |                     curexc_type
```
```
[balay@pj01 spack]$ ./bin/spack find
==> 15 installed packages
-- linux-fedora31-skylake / gcc@9.3.1 ---------------------------
arpack-ng@3.8.0  hypre@2.20.0  mpich@3.3            parmetis@4.0.3  py-mpi4py@3.0.3  py-petsc4py@3.15.0  slepc@3.15.0        zlib@1.2.11
hdf5@1.10.7      metis@5.1.0   netlib-lapack@3.8.0  petsc@3.15.0    py-numpy@1.17.4  python@3.7.9        superlu-dist@6.4.0
```
[spack-build-out.txt](https://github.com/spack/spack/files/6244162/spack-build-out.txt)

--- a/var/spack/repos/builtin/packages/py-slepc4py/ldshared-dev.patch
+++ b/var/spack/repos/builtin/packages/py-slepc4py/ldshared-dev.patch
@@ -1,0 +1,13 @@
+diff --git a/src/binding/slepc4py/conf/baseconf.py b/src/binding/slepc4py/conf/baseconf.py
+index b0707a65a..44ba8f194 100644
+--- a/src/binding/slepc4py/conf/baseconf.py
++++ b/src/binding/slepc4py/conf/baseconf.py
+@@ -213,7 +213,7 @@ class PetscConfig:
+         ldshared = getenv('LDSHARED', ldshared)
+         ldflags = getenv('LDFLAGS', cflags + ' ' + (ldflags or ''))
+         ldcmd = split_quoted(ld) + split_quoted(ldflags)
+-        ldshared = [flg for flg in split_quoted(ldshared) if flg not in ldcmd]
++        ldshared = [flg for flg in split_quoted(ldshared) if flg not in ldcmd and (flg.find('/lib/spack/env')<0)]
+         ldshared = str.join(' ', ldshared)
+         #
+         def get_flags(cmd):

--- a/var/spack/repos/builtin/packages/py-slepc4py/ldshared.patch
+++ b/var/spack/repos/builtin/packages/py-slepc4py/ldshared.patch
@@ -1,0 +1,13 @@
+diff --git a/src/binding/slepc4py/conf/baseconf.py b/src/binding/slepc4py/conf/baseconf.py
+index b0707a65a..44ba8f194 100644
+--- a/conf/baseconf.py
++++ b/conf/baseconf.py
+@@ -213,7 +213,7 @@ class PetscConfig:
+         ldshared = getenv('LDSHARED', ldshared)
+         ldflags = getenv('LDFLAGS', cflags + ' ' + (ldflags or ''))
+         ldcmd = split_quoted(ld) + split_quoted(ldflags)
+-        ldshared = [flg for flg in split_quoted(ldshared) if flg not in ldcmd]
++        ldshared = [flg for flg in split_quoted(ldshared) if flg not in ldcmd and (flg.find('/lib/spack/env')<0)]
+         ldshared = str.join(' ', ldshared)
+         #
+         def get_flags(cmd):

--- a/var/spack/repos/builtin/packages/py-slepc4py/package.py
+++ b/var/spack/repos/builtin/packages/py-slepc4py/package.py
@@ -16,6 +16,7 @@ class PySlepc4py(PythonPackage):
 
     maintainers = ['dalcinl', 'joseeroman', 'balay']
 
+    version('main', branch='main')
     version('3.15.0', sha256='b1948a325f61e7acebc8a79eae4f31991dd1d4eb508437aced1085c4f76918ce')
     version('3.13.0', sha256='780eff0eea1a5217642d23cd563786ef22df27e1d772a1b0bb4ccc5701df5ea5')
     version('3.12.0', sha256='d8c06953b7d00f529a9a7fd016dfa8efdf1d05995baeea7688d1d59611f424f7')
@@ -25,6 +26,10 @@ class PySlepc4py(PythonPackage):
     version('3.8.0',  sha256='988815b3650b69373be9abbf2355df512dfd200aa74b1785b50a484d6dfee971')
     version('3.7.0',  sha256='139f8bb325dad00a0e8dbe5b3e054050c82547936c1b6e7812fb1a3171c9ad0b')
 
+    patch('ldshared.patch', when='@:99')
+    patch('ldshared-dev.patch', when='@main')
+
+    depends_on('py-cython', type='build', when='@main')
     depends_on('py-setuptools', type='build')
 
     depends_on('py-petsc4py', type=('build', 'run'))
@@ -48,3 +53,11 @@ class PySlepc4py(PythonPackage):
     depends_on('slepc@3.8:3.8.99', when='@3.8:3.8.99')
     depends_on('slepc@3.7:3.7.99', when='@3.7:3.7.99')
     depends_on('slepc@3.6:3.6.99', when='@3.6:3.6.99')
+
+    @property
+    def build_directory(self):
+        import os
+        if self.spec.satisfies('@main'):
+            return os.path.join(self.stage.source_path, 'src', 'binding', 'slepc4py')
+        else:
+            return self.stage.source_path

--- a/var/spack/repos/builtin/packages/py-slepc4py/package.py
+++ b/var/spack/repos/builtin/packages/py-slepc4py/package.py
@@ -17,7 +17,7 @@ class PySlepc4py(PythonPackage):
     maintainers = ['dalcinl', 'joseeroman', 'balay']
 
     version('main', branch='main')
-    version('3.15.0', sha256='b1948a325f61e7acebc8a79eae4f31991dd1d4eb508437aced1085c4f76918ce')
+    version('3.15.0', sha256='dd190b8dfd21cc69ac1d208e0d6784fc3a3141d5c4a0e433af26e20bfa5e917a')
     version('3.13.0', sha256='780eff0eea1a5217642d23cd563786ef22df27e1d772a1b0bb4ccc5701df5ea5')
     version('3.12.0', sha256='d8c06953b7d00f529a9a7fd016dfa8efdf1d05995baeea7688d1d59611f424f7')
     version('3.11.0', sha256='1e591056beee209f585cd781e5fe88174cd2a61215716a71d9eaaf9411b6a775')

